docs/source/notes/modules.rst

```diff
@@ -0,0 +1,433 @@
+This module applies an affine transformation to its input.
+
+.. code-block:: python
+
+   import torch
+   from torch import nn
+
+   class MyLinear(nn.Module):
+     def __init__(self, in_features, out_features):
+       super().__init__()
+       self.weight = nn.Parameter(torch.randn(in_features, out_features))
+       self.bias = nn.Parameter(torch.randn(out_features))
+
+     def forward(self, input):
+       return (input @ self.weight) + self.bias
+
+This simple module has the following fundamental characteristics of modules:
+
+* **It inherits from the base Module class.**
+  All modules should subclass :class:`~torch.nn.Module` for composability with other modules.
+* **It defines some "state" that is used in computation.**
+  Here, the state consists of randomly-initialized ``weight`` and ``bias`` tensors that define the affine
+  transformation. Because each of these is defined as a :class:`~torch.nn.parameter.Parameter`, the module
+  will automatically track them and return them from calls to :func:`~torch.nn.Module.parameters`. Parameters can be
+  considered the "learnable" aspects of the module's computation (more on this later). Note that modules
+  are not *required* to have state, and can also be stateless.
+* **It defines a forward() function that performs the computation.** For this affine transformation module, the input
+  is matrix-multiplied with the ``weight`` parameter (using the ``@`` short-hand notation) and added to the ``bias``
+  parameter to produce the output. More generally, the ``forward()`` implementation for a module can perform arbitrary
+  computation involving any number of inputs and outputs.
+
+This simple module demonstrates how modules package state and computation. Instances of this module can be
+constructed and called:
+
+.. code-block:: python
+
+   m = MyLinear(4, 3)
+   sample_input = torch.randn(4)
+   m(sample_input)
+   : tensor([-0.3037, -1.0413, -4.2057], grad_fn=<AddBackward0>)
+
+Note that the module itself is callable, and that calling it calls its ``forward()`` function.
+The full set of parameters registered by the module can be iterated through via a call to
+:func:`~torch.nn.Module.parameters` or :func:`~torch.nn.Module.named_parameters`,
+where the latter includes each parameter's name:
+
+.. code-block:: python
+
+   for parameter in m.named_parameters():
+     print(parameter)
+   : ('weight', Parameter containing:
+   tensor([[ 1.0597,  1.1796,  0.8247],
+           [-0.5080, -1.2635, -1.1045],
+           [ 0.0593,  0.2469, -1.4299],
+           [-0.4926, -0.5457,  0.4793]], requires_grad=True))
+   ('bias', Parameter containing:
+   tensor([ 0.3634,  0.2015, -0.8525], requires_grad=True))
+
+In general, the parameters registered by a module are aspects of the module's computation that should be
+"learned". A later section of this note shows how to update these parameters using one of PyTorch's optimizers.
```

Comment:
I think Optimizers should be capitalized throughout this.

Reply:
Sounds good- should we be consistent by capitalizing Modules throughout as well?